### ISSUE TITLE
Deep Tree Echo Reservoir-Orchestrated Training — Evolution Plan

### ISSUE THREAD

**Issue by drzo** [9cog/echoself#57]
<div class="markdown-body MarkdownRenderer-module__container__sJ8jf" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><h1>Deep Tree Echo Reservoir-Orchestrated Training — Evolution Plan</h1>
<h2>Guiding principle (from your stated objective)</h2>
<p>Tokenization, topology, and model size become <strong>dynamic, persona-driven configuration</strong> — not fixed GPT-2 constants. The transformer is <em>emulated between reservoirs &amp; ridges</em>: an Echo State Network orchestrates the training loop, and ridge-regression readouts mediate between the reservoir and the transformer.</p>
<h2>Phase 0 — Groundwork &amp; safety (no behavior change)</h2>
<ul>
<li node="[object Object]">Add a feature-flag/config seam (<code node="[object Object]">reservoir_mode: off|shadow|orchestrated</code>, default <code node="[object Object]">off</code>) in <code node="[object Object]">nanecho_config.json</code> / <code node="[object Object]">TrainingConfig</code> so all new machinery is opt-in and current GPT-2 CI training keeps passing.</li>
<li node="[object Object]">Extract today's hardcoded GPT-2 constants (in <code node="[object Object]">NanEcho/runtime.py</code>, <code node="[object Object]">train_nanecho.py</code>, checkpoint provenance) into a single <code node="[object Object]">TokenizerSpec</code> / <code node="[object Object]">ModelSpec</code> dataclass — the GPT-2 values become <em>one instance</em> of a spec, not the law. Keep the existing provenance validator working by having it validate against the <em>active</em> spec rather than hardcoded GPT-2 (GPT-2 remains the default spec until a persona-fit one replaces it).</li>
<li node="[object Object]">Add a regression test pinning current behavior with the flag off.</li>
</ul>
<h2>Phase 1 — Dynamic tokenization (persona-gripped)</h2>
<ul>
<li node="[object Object]">Introduce a <code node="[object Object]">TokenizerAdapter</code> interface (<code node="[object Object]">name, vocab_size, eos_token, eos_token_id, encode, decode, provenance()</code>) so <code node="[object Object]">NanEchoTokenizer</code>, <code node="[object Object]">dte_tokenizer/</code> (the existing custom tokenizer assets), and char/BPE variants are interchangeable.</li>
<li node="[object Object]">Implement a <strong>tokenizer search</strong>: train candidate tokenizers (vocab sizes spanning e.g. 256–50k) on the persona corpus (<code node="[object Object]">NanEcho/persona_corpus/</code>, conversation patterns), then score each candidate with a <em>grip metric</em> built from existing signals: <code node="[object Object]">score_persona_text</code> coverage on tokenizer-round-tripped persona text, token fertility on persona samples, and downstream perplexity of a tiny probe model.</li>
<li node="[object Object]">Persist the winning tokenizer's provenance into dataset <code node="[object Object]">metadata.json</code> (extending the schema I fixed in the workflows), so the provenance validator passes for <em>any</em> persona-selected tokenizer — not just GPT-2.</li>
</ul>
<h2>Phase 2 — Reservoir layer around the transformer</h2>
<ul>
<li node="[object Object]">Promote <code node="[object Object]">NanEcho/dte_nodes/EchoReservoir</code> into the training path (currently only used in the standalone <code node="[object Object]">dte_esn_pipeline.py</code> demo):
<ul>
<li node="[object Object]"><strong>Input reservoir</strong> encodes token sequences into rich dynamical states (replacing/extending the transformer's static embedding on the persona axis).</li>
<li node="[object Object]"><strong>Ridge readouts</strong> (<code node="[object Object]">CognitiveReadout</code> — ridge regression) sit between reservoir states and the transformer's logits/embeddings: the "transformer emulated between reservoirs &amp; ridges."</li>
<li node="[object Object]"><strong>AARRelation feedback loop</strong> feeds persona alignment back into reservoir driving.</li>
</ul>
</li>
<li node="[object Object]"><code node="[object Object]">NanEchoModel</code> gains an optional <code node="[object Object]">ReservoirWrapper</code> that routes embedding → reservoir → ridge → block stack, controlled by the Phase-0 flag. Transformer blocks (AdaptiveAttention, PersonaDimension, RecursiveReasoning, HypergraphPatternEncoder) remain; the reservoir modulates their inputs.</li>
</ul>
<h2>Phase 3 — ESN orchestration of training (the conductor)</h2>
<ul>
<li node="[object Object]">Create <code node="[object Object]">ReservoirOrchestrator</code> (extending <code node="[object Object]">train_cached.CachedNanEchoTrainer</code>) that lets the ESN <em>drive training hyperparameters dynamically</em>, replacing fixed schedules:
<ul>
<li node="[object Object]">Reservoir state statistics (spectral radius proxy, state entropy, Lyapunov-style stability) + persona grip scores feed an ESN whose ridge readout outputs <strong>per-interval decisions</strong>: connection growth rate (already incremental in NanEcho), effective LR scaling, recursion depth (<code node="[object Object]">min/max_recursion_depth</code>), and persona dimension weights.</li>
<li node="[object Object]">Ridge readout is re-fit online every eval interval from (state, decision) → (grip improvement) pairs — ridge regression as the "ridge" between reservoir and transformer updates.</li>
</ul>
</li>
<li node="[object Object]">Checkpoint the orchestrator state alongside model checkpoints (extend <code node="[object Object]">training_cache.py</code> metadata), so cumulative training resumes with its conductor intact.</li>
</ul>
<h2>Phase 4 — Dynamic topology &amp; model size</h2>
<ul>
<li node="[object Object]">Extend the existing <code node="[object Object]">ConnectionMask</code> growth mechanism into <strong>topology search</strong>: reservoir readout proposes which persona dimensions/attention heads/layers to grow or prune based on measured grip contribution per component (using <code node="[object Object]">evaluate_echo_self_quality</code> per-dimension scores from <code node="[object Object]">train_nanecho.py</code>).</li>
<li node="[object Object]">Add <strong>model-size adaptation</strong>: train a sweep of small configs (n_layer/n_embd grid), let the orchestrator pick the smallest config whose grip score saturates (elastic grip curve), and support warm-start growing from smaller checkpoints (NanEcho's iterative connection building already provides the growth substrate).</li>
<li node="[object Object]">Persist the selected topology in checkpoint metadata so <code node="[object Object]">NanEchoRuntime.load</code> reconstructs the dynamic shape (the runtime already validates persona dimensions from state dicts — extend this to full topology).</li>
</ul>
<h2>Phase 5 — Grip evaluation &amp; CI integration</h2>
<ul>
<li node="[object Object]">Build <code node="[object Object]">evaluation/grip_benchmark.py</code>: persona-conversation-pattern suite scoring tokenizer+topology+size combos end-to-end; becomes the objective function for Phases 1 &amp; 4.</li>
<li node="[object Object]">CI: add a <code node="[object Object]">grip-eval</code> workflow job that runs the benchmark on trained checkpoints and uploads grip reports as artifacts alongside existing training summaries.</li>
<li node="[object Object]">Update the workflow fallback data generators (the ones I fixed) to emit tokenizer provenance for whichever spec Phase 1 selects, keeping CI green throughout.</li>
</ul>
<h2>Phase 6 — Migration &amp; cleanup</h2>
<ul>
<li node="[object Object]">Once reservoir-orchestrated runs match/exceed baseline grip on the persona corpus, flip the default <code node="[object Object]">reservoir_mode</code> to <code node="[object Object]">orchestrated</code> for persona training while keeping <code node="[object Object]">off</code> for compatibility.</li>
<li node="[object Object]">Deprecate GPT-2-only assumptions in docs (<code node="[object Object]">CLAUDE.MD</code>, <code node="[object Object]">NANECHO_README.md</code>) and record the new dynamic-spec contract.</li>
</ul>
<h2>Verification strategy throughout</h2>
<ul>
<li node="[object Object]"><code node="[object Object]">python3 -m pytest --ignore=test_data_validation.py --ignore=NanEcho/test_production_runtime.py</code> (per repo test conventions) after each phase</li>
<li node="[object Object]"><code node="[object Object]">npm run lint &amp;&amp; npm run typecheck &amp;&amp; npm run build</code> untouched but run to catch accidental breakage</li>
<li node="[object Object]">CI-mode training run (<code node="[object Object]">--echo_depth=3</code>, small iters) with flag on and off as the integration gate per phase</li>
</ul>
<h2>Sequencing rationale</h2>
<p>Phase 0–1 deliver immediate value (dynamic tokenization) with low risk; Phases 2–3 realize the core reservoir-orchestrated vision reusing the existing <code node="[object Object]">dte_nodes</code> rather than building from scratch; Phases 4–6 make topology/size dynamic and productionize it.</p></div>
